File: bindings/csharp/src/TonMnemonic.cs
//
// Copyright (c) Fela Ameghino 2026
//
// Distributed under the MIT License. (See accompanying file LICENSE or copy at
// https://opensource.org/licenses/MIT)
//
// The values the facade hands back that are not walletkit models.
//
// Hand-written: a mnemonic is not a string (confusing it with one is how secrets
// end up in logs), and the creation result is just a pair.
//
using System.Collections.Generic;

namespace Ton.WalletKit
{
    /// <summary>A generated mnemonic and the wallet built from it (kit-ios TONWalletCreationResult).</summary>
    public sealed class TonWalletCreationResult
    {
        /// <summary>The words. Persist these: walletkit stores no keys.</summary>
        public TonMnemonic Mnemonic { get; }

        /// <summary>The adapter derived from them; register it with AddAsync.</summary>
        public ITonWalletAdapter WalletAdapter { get; }

        /// <summary>Pairs a mnemonic with the adapter derived from it.</summary>
        public TonWalletCreationResult(TonMnemonic mnemonic, ITonWalletAdapter walletAdapter)
        {
            Mnemonic = mnemonic;
            WalletAdapter = walletAdapter;
        }
    }

    /// <summary>
    /// A TON mnemonic (kit-ios TONMnemonic).
    ///
    /// This is the only thing that can rebuild a wallet, and the core never
    /// persists it — storing it securely is the app's job.
    /// </summary>
    public sealed class TonMnemonic
    {
        /// <summary>The words, in order.</summary>
        public IReadOnlyList<string> Value { get; }

        /// <summary>Wraps a word list.</summary>
        public TonMnemonic(IReadOnlyList<string> value)
        {
            Value = value;
        }

        /// <summary>The words separated by spaces, the usual storage form.</summary>
        public override string ToString()
        {
            return Value == null ? string.Empty : string.Join(" ", Value);
        }

        /// <summary>Parses a space-separated mnemonic.</summary>
        public static TonMnemonic Parse(string value)
        {
            return new TonMnemonic(value == null
                ? new string[0]
                : value.Split(new[] { ' ', '\t', '\n', '\r' }, System.StringSplitOptions.RemoveEmptyEntries));
        }
    }
}

File: bindings/android/src/main/java/org/ton/walletkit/core/TonMnemonic.java
//
// Copyright (c) Fela Ameghino 2026
//
// Distributed under the MIT License. (See accompanying file LICENSE or copy at
// https://opensource.org/licenses/MIT)
//

package org.ton.walletkit.core;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.List;

/**
 * A TON mnemonic.
 *
 * <p>Wrapped rather than passed as a String so it cannot be mistaken for prose
 * and logged. It is the only thing that can rebuild a wallet, and the core never
 * persists it — storing it securely is the app's job.
 */
public final class TonMnemonic {

    private final List<String> value;

    public TonMnemonic(List<String> value) {
        this.value = Collections.unmodifiableList(new ArrayList<String>(value));
    }

    public static TonMnemonic parse(String value) {
        return new TonMnemonic(Arrays.asList(value.trim().split("\s+")));
    }

    public List<String> getValue() {
        return value;
    }

    /** The words separated by spaces — the usual storage form. Never log this. */
    public String joined() {
        StringBuilder out = new StringBuilder();
        for (String word : value) {
            out.append(out.length() == 0 ? "" : " ").append(word);
        }
        return out.toString();
    }

    /** Deliberately not the words: toString ends up in logs. */
    @Override
    public String toString() {
        return "TonMnemonic(" + value.size() + " words)";
    }
}
